src/base/Identifier.ml

```diff
@@ -19,21 +19,89 @@
 open Core_kernel
 open ErrorUtils
 
-type 'rep t = Ident of string * 'rep [@@deriving sexp]
+module type QualifiedName = sig
 
-let asId i = Ident (i, dummy_loc)
+  type t [@@deriving sexp]
 
-let asIdL i loc = Ident (i, loc)
+  val as_string : t -> string
 
-let get_id i = match i with Ident (x, _) -> x
+  val as_error_string : t -> string
 
-let get_rep i = match i with Ident (_, l) -> l
+  val equal_id : t -> t -> bool
+  
+  val compare_id : t -> t -> int
+    
+end
+
+module FlattenedName = struct
+
+  type t = string [@@deriving sexp]
+
+  let as_string n = n
+
+  let as_error_string n = n
+
+  let equal_id = String.(=)
+
+  let compare_id = String.compare
+
+end
+
+
+module type ScillaIdentifier = sig
+
+  module Name : QualifiedName
+    
+  type 'rep t = Ident of Name.t * 'rep [@@deriving sexp]
+
+  val asId : Name.t -> loc t
+
+  val asIdL : Name.t -> 'a -> 'a t
+
+  val get_id : 'a t -> Name.t
+
+  val get_rep : 'a t -> 'a
+
+  val as_string : 'a t -> string
+
+  val as_error_string : 'a t -> string
+
+  (* A few utilities on id. *)
+  val equal_id : 'a t -> 'b t -> bool
+
+  val compare_id : 'a t -> 'b t -> int
+
+  val dedup_id_list : 'a t list -> 'a t list
+
+  val is_mem_id : 'a t -> 'a t list -> bool
```

Comment:
Looks like implementations of these two should be the same in all instantiations.
If this is true, then we should move them outside the structure.

Reply:
Yeah, but we'll only have one functor anyway. I'll move it outside.

Reply:
The implementation is the same for all instantiations, but the type is not, becaue the type contains `ScillaIdentifier.t`. Therefore, the functions need to be inside the structure.